src/bytes.rs

```diff
@@ -1101,6 +1101,16 @@ unsafe fn release_shared(ptr: *mut Shared) {
     Box::from_raw(ptr);
 }
 
+fn ptr_map<F>(ptr: *mut u8, f: F) -> *mut u8
+where
+    F: FnOnce(usize) -> usize,
+{
+    let old_addr = ptr as usize;
+    let new_addr = f(old_addr);
+    let diff = new_addr.wrapping_sub(old_addr);
+    ptr.wrapping_add(diff)
+}
```

Comment:
Unfortunately, it seems like this doesn't really optimize as well as I had hoped. You can view godbolt [here][1] to see the difference.

[1]: https://rust.godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:rust,selection:(endColumn:2,endLineNumber:12,positionColumn:2,positionLineNumber:12,selectionStartColumn:2,selectionStartLineNumber:12,startColumn:2,startLineNumber:12),source:'pub+fn+test(a:+usize)+-%3E+usize+%7B%0A++++let+new+%3D+a+%7C+1%3B%0A++++let+diff+%3D+new.wrapping_sub(a)%3B%0A++++a.wrapping_add(diff)%0A%7D%0A%0Apub+fn+test2(a:+*mut+u8)+-%3E+*mut+u8+%7B%0A++++let+a_val+%3D+a+as+usize%3B%0A++++let+new+%3D+a_val+%7C+1%3B%0A++++let+diff+%3D+new.wrapping_sub(a_val)%3B%0A++++a.wrapping_add(diff)%0A%7D'),l:'5',n:'0',o:'Rust+source+%231',t:'0')),k:49.99999999999999,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:nightly,filters:(b:'0',binary:'1',commentOnly:'0',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'1',trim:'1'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:rust,libs:!(),options:'-C+opt-level%3D3',selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1,tree:'1'),l:'5',n:'0',o:'rustc+nightly+(Rust,+Editor+%231,+Compiler+%231)',t:'0')),k:49.99999999999999,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4

Reply:
hmm, that's kind of sad...

Reply:
One option is to go through a normal pointer cast for non-miri release mode.

Reply:
Is this valid? This seems to optimize well, though I have no idea why. I just find this pattern for writing pointer packing more intuitive.
https://rust.godbolt.org/z/nK8M8hshM